src/common/media.cpp

```diff
@@ -241,45 +247,97 @@ float media_print_get_percent_done(void) {
+            *state = GCodeFilter::State::Eof;
+        } else {
+            *state = GCodeFilter::State::Error;
+        }
+    }
+
+    *state = GCodeFilter::State::Ok;
+
+    media_current_position++;
+    media_loop_read++;
+
+    return byte;
+}
+
 void media_loop(void) {
+    if (media_print_state == media_print_state_PAUSING) {
+        f_close(&media_print_fil);
+        int index_r = queue.index_r;
+        media_gcode_position = media_current_position = media_queue_position[index_r];
+        queue.clear();
+        media_print_state = media_print_state_PAUSED;
+        return;
+    }
+
     _usbhost_reenum();
-    if (media_print_state == media_print_state_PRINTING) {
-        char buffer[MAX_CMD_SIZE + 1];
-        char *pch;
-        if (!f_eof(&media_print_fil))              //check eof
-            while (queue.length < (BUFSIZE - 1)) { //keep one free slot for serial commands
-                if (f_gets(buffer, MAX_CMD_SIZE, &media_print_fil)) {
-                    pch = strchr(buffer, '\r');
-                    if (pch)
-                        *pch = 0; //replace CR with 0
-                    pch = strchr(buffer, '\n');
-                    if (pch)
-                        *pch = 0; //replace LF with 0
-                    pch = strchr(buffer, ';');
-                    if (pch)
-                        *pch = 0;         //replace ; with 0 (cut comment)
-                    if (strlen(buffer)) { //enqueue only not empty lines
-                        queue.enqueue_one(buffer, false);
-                        int index_w = queue.index_r + queue.length - 1; //calculate index_w because it is private
-                        if (index_w >= BUFSIZE)
-                            index_w -= BUFSIZE;                                 //..
-                        media_queue_position[index_w] = media_current_position; //save current position
-                        media_queue_line[index_w] = media_current_line;         //save current line
-                    }
-                    media_current_position = f_tell(&media_print_fil); //update current position
-                    media_current_line++;                              //update current line
-                } else {
-                    if (f_eof(&media_print_fil)) //we need check eof also after read operation
-                        media_print_stop();      //stop on eof
-                    else {
-                        set_warning(WarningType::USBFlashDiskError);
-                        media_print_pause(); //pause in other case (read error - media removed)
-                    }
-                    break;
-                }
+
+    if (media_print_state != media_print_state_PRINTING) {
+        return;
+    }
+
+    media_loop_read = 0;
+
+    while (queue.length < (BUFSIZE - 1)) { // Keep one free slot for serial commands
+        GCodeFilter::State state;
+        char *gcode = gcode_filter.nextGcode(&state);
+
+        if (state == GCodeFilter::State::Skip) {
```

Comment:
`switch`?

Reply:
It's handling just two cases out of four possible. Third state is checked only under specified condition. I don't think switch would increase readability in this case.